Add `no_std` support to `wgpu-types`

bushrat011899 opened the pull request:
**Connections**
- Contributes to #6826

**Description**
For a possible `no_std` version of `wgpu` (as requested in the above issue), a good first step would be for the already slim `wgpu-types` crate to be `no_std` compatible. This is what this PR accomplishes.

First, I disabled default features on the crates `wgpu-types` relies on (disabling their respective `std` features). As these dependencies are managed through the workspace, they had to be disabled there, requiring re-enabling `default` in the adjacent crates.

Next, I added `std` and `default` features to `wgpu-types`. This ensures users who do not care about `no_std` compatibility have minimal friction when using this crate. I also added the following lints to `wgpu-types`:

```toml
[lints.clippy]
std_instead_of_core = "warn"
std_instead_of_alloc = "warn"
alloc_instead_of_core = "warn"
```

These are very helpful when working on a crate with features for `std` and `alloc`, since they will inform you if you're being overly restrictive on how you import items which are re-exported.

Next, I added the following to the `lib.rs` file:

```rust
#![no_std]

#[cfg(feature = "std")]
extern crate std;

extern crate alloc;
```

By being unconditionally `no_std` and optionally including `std` through a feature, this crate will always use the `core` implicit prelude instead of switching between `core` and `std` based on the `std` feature. This was a headache in Bevy's `no_std` efforts.

Finally, I added appropriate feature gates where required, and tweaked one particular method for serializing `Astc` texture formats ~~to ensure the `serde` feature didn't require the `alloc` one. There's nothing wrong with requiring `alloc` for `serde`, but it was a minor change to decouple these two features.~~ ~~Change isn't required but I've already written it, and it does avoid allocation, so it's marginally more performant.~~ Reverted.

**Testing**
Tested through compiling `wgpu-types` using the following commands:
```bash
cargo check -p wgpu-types --no-default-features --target x86_64-unknown-none
cargo check -p wgpu-types --no-default-features --target x86_64-unknown-none --features strict_asserts,fragile-send-sync-non-atomic-wasm,serde,counters
```

<!--
Thanks for filing! The codeowners file will automatically request reviews from the appropriate teams.

After you get a review and have addressed any comments, please explicitly re-request a review from the
person(s) who reviewed your changes. This will make sure it gets re-added to their review queue - you're no bothering us!
-->

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `taplo format`.
- [x] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
  - [ ] `--target wasm32-unknown-emscripten`
- [ ] Run `cargo xtask test` to run tests.
- [x] Add change to `CHANGELOG.md`. See simple instructions inside file.

**Notes**

~~I have made the assumption that conditional `alloc` may provide utility to this crate, but that may be faulty. It could be removed and the `extern crate alloc` line could always be included.~~

bushrat011899 committed:
Initial Commit

bushrat011899 committed:
Remove now-redundant `format` import

bushrat011899 committed:
Update CHANGELOG.md

bushrat011899 committed:
Appropriately feature-gate `texture_format_serialize` test

cwfitzgerald commented:
Just one note while you're working on this, I don't think we need no-alloc support; the rest of the stack depends on it anyway.

bushrat011899 committed:
Remove `alloc` feature

Also fixed some documentation links and a Wasm `std` import

bushrat011899 commented:
> Just one note while you're working on this, I don't think we need no-alloc support; the rest of the stack depends on it anyway.

Easy, I've removed that feature and fixed (hopefully) the last CI issue too.

cwfitzgerald commented on wgpu-types/src/lib.rs:
```diff
@@ -3056,33 +3116,7 @@ impl Serialize for TextureFormat {
-                s = format!("astc-{block}-{channel}");
-                &s
-            }
+            TextureFormat::Astc { block, channel } => astc_texture_format_to_str(block, channel),
```
I think this change can be rolled back now that we're back on alloc unconditionally

bushrat011899 replied:
Agreed. Rolled back to minimise changes.

cwfitzgerald commented on wgpu-types/src/lib.rs:
```diff
@@ -7,15 +7,24 @@
+use alloc::{string::String, vec, vec::Vec};
+use core::hash::{Hash, Hasher};
+use core::mem::size_of;
+use core::{num::NonZeroU32, ops::Range};
```
Could we keep these in a single use? I realize they weren't before, but we should be aggressively combining

bushrat011899 replied:
That's my preferred style too. Done.

cwfitzgerald commented on wgpu-types/src/lib.rs:
```diff
@@ -7688,6 +7724,7 @@ impl Default for ShaderRuntimeChecks {
 ///
 /// `wgpu::utils::init::dx12_shader_compiler_from_env` can be used to set the compiler
 /// from the `WGPU_DX12_SHADER_COMPILER` environment variable, but this should only be used for testing.
+#[cfg(feature = "std")]
```
Lets change these from a PathBuf to a String - the api would need to change to allow this to be std-only and that is very weird as the idea is you always pass these settings on all platforms.

bushrat011899 replied:
Agreed. I did a similar thing in Bevy for its `no_std` support too.

cwfitzgerald commented on wgpu-types/src/lib.rs:
```diff
@@ -7736,6 +7773,7 @@ pub enum Gles3MinorVersion {
 }
 
 /// Options for creating an instance.
+#[cfg(feature = "std")]
```
Yeah this won't work, see the note about pathbuf.

cwfitzgerald commented on wgpu-types/src/lib.rs:
```diff
@@ -7,15 +7,24 @@
+
+extern crate alloc;
+
+use alloc::{string::String, vec, vec::Vec};
```
What's this `vec` import about.

bushrat011899 replied:
I'll double check, but the pattern is:

```rust
use alloc::vec; // Gets the vec![] macro
use alloc::vec::Vec; // Gets the Vec type
```

I believe `vec!` was being used, but I'll double-check that.

To preempt, you can't combine them any smaller than `alloc::{vec, vec::Vec}`. For example, `alloc::vec::{self, Vec}` imports the module `vec` and the type `Vec`, not the macro and type.

cwfitzgerald replied:
That's wacky weird

cwfitzgerald commented on wgpu-types/src/lib.rs:
```diff
@@ -1076,6 +1085,7 @@ impl InstanceFlags {
     /// The environment variables are named after the flags prefixed with "WGPU_". For example:
     /// - WGPU_DEBUG
     /// - WGPU_VALIDATION
+    #[cfg(feature = "std")]
```
This should be a no-op - basically as-if no env vars are set. Maybe this could be done through a cfg on env.

bushrat011899 replied:
Makes sense, I'll look at changing it around.

bushrat011899 replied:
Ended up moving the `cfg(...)` inside the `env` function as discussed.

cwfitzgerald commented on wgpu-types/src/lib.rs:
```diff
@@ -1076,6 +1085,7 @@ impl InstanceFlags {
     /// The environment variables are named after the flags prefixed with "WGPU_". For example:
     /// - WGPU_DEBUG
     /// - WGPU_VALIDATION
+    #[cfg(feature = "std")]
```
Same comment here re env

cwfitzgerald reviewed (changes requested):
Got some comments, but generally looking good.

The other thing that needs to be done is get CI support for a no-std platform so we don't break no-std.

Lets do testing on `wasm32v1-none` as that was the inciting incident for this work.

You can add another variant after https://github.com/gfx-rs/wgpu/blob/trunk/.github/workflows/ci.yml#L145-L148 which is the wasmv1 target. It will be a new kind called `no-std` and for now lets use the nightly toolchain, so add `toolchain: nightly` (not required but this makes setup easier)

Then add a new block similar to https://github.com/gfx-rs/wgpu/blob/trunk/.github/workflows/ci.yml#L234-L248 which is conditioned on `if: matrix.kind == "no-std"` and contains the exact things you were testing with. We can expand them as more of wgpu gets no-std access.

bushrat011899 committed:
Revert change to `Serialize` for `TextureFormat`

bushrat011899 committed:
Combine use statements

bushrat011899 committed:
Switch from `PathBuf` to `String`

bushrat011899 committed:
Consider environmental flags as unset on `no_std`

bushrat011899 commented:
Noting here that there is a bug in the current version of `web-sys` (0.3.76) which prevents compilation on the new `wasm32v1-none` platform. It has been resolved in [this PR](https://github.com/rustwasm/wasm-bindgen/pull/4378) which is already merged. Until a new version of `web-sys` is published, `wgpu-types` will not compile on `wasm32v1-none`. However, by removing the `ImageData` feature (the one in `web-sys` that causes the current compiler error) I can see compilation fails within `wgpu-types` on `web_sys::ImageData` being missing. This indicates to me that this _will_ successfully compile once `web-sys` publishes their next version.

bushrat011899 committed:
Fix missing `format!`

bushrat011899 committed:
Add new CI tasks for `no_std` testing

bushrat011899 commented:
Ok I've addressed all the items from that review (thanks for being so thorough btw!). I've also added a new `no_std` kind of CI task with _2_ matrix entries; one for bare-metal x86 (`x86_64-unknown-none`) and one for `wasm32v1-none`. As mentioned above, `wasm32v1-none` will fail until `web-sys` updates, so I included bare-metal x86 so we can verify that the CI task works as expected.

cwfitzgerald commented on .github/workflows/ci.yml:
```diff
@@ -151,6 +151,17 @@ jobs:
             target: wasm32-unknown-emscripten
             kind: wgpu-only
+          
+          - name: WebAssembly Core 1.0
+            os: ubuntu-22.04
+            target: wasm32v1-none
+            kind: no-std
+          
+          # Bare-metal x86-64
+          - name: Bare-metal x86_64
+            os: ubuntu-22.04
+            target: x86_64-unknown-none
+            kind: no-std
```
Because wasm32v1-none can't compile, let's comment it out, with a note that includes a link to the issue. Additionally let's note why we're testing bare metal x64, an odd platform

bushrat011899 replied:
Added some notes and linked to the relevant PR describing the issue with `web-sys`.

cwfitzgerald commented on .github/workflows/ci.yml:
```diff
@@ -151,6 +151,17 @@ jobs:
+          - name: WebAssembly Core 1.0
+            os: ubuntu-22.04
+            target: wasm32v1-none
+            kind: no-std
```
Need to add toolchain: nightly just like visionos

bushrat011899 replied:
Whoops forgot that one, fixed!

cwfitzgerald reviewed (changes requested):
Looking good. Have some notes about CI, and CI is mad.

bushrat011899 committed:
Comment out known failing CI matrix option

bushrat011899 committed:
Update all usage of `Dx12Compiler::DynamicDxc`

bushrat011899 committed:
Added comments to CI

bushrat011899 commented:
Addressed the latest round of comments and fixed that CI issue (forgot to cascade the change in `Dx12Compiler::DynamicDxc` to all crates)

bushrat011899 committed:
Merge remote-tracking branch 'upstream/trunk' into no_std-wgpu-types

cwfitzgerald commented on .github/workflows/ci.yml:
```diff
@@ -246,6 +263,19 @@ jobs:
+        
+      # Building for no_std platforms where every feature is enabled except "std".
+      - name: check no_std
+        if: matrix.kind == 'no_std'
```
```suggestion
        if: matrix.kind == 'no-std'
```

Used different strings for no-std

bushrat011899 replied:
Ah silly mistake on my part, thanks for fixing that!

cwfitzgerald commented on .github/workflows/ci.yml:
```diff
@@ -151,6 +151,23 @@ jobs:
+          # Bare-metal x86-64
+          # TODO: Remove once web-sys updates past 0.3.76
+          # Included while wasm32v1-none is failing to ensure `no_std` does not regress
+          - name: Bare-metal x86_64
```
```suggestion
          - name: no-std x86_64
```

cwfitzgerald committed:
Update .github/workflows/ci.yml

cwfitzgerald committed:
Update .github/workflows/ci.yml

cwfitzgerald commented:
Bleh CI crap, I'll deal with this

cwfitzgerald commented on .github/workflows/ci.yml:
```diff
@@ -151,6 +151,23 @@ jobs:
+            os: ubuntu-22.04
+            target: x86_64-unknown-none
+            kind: no-std
+            toolchain: nightly
```
```suggestion
```

cwfitzgerald committed:
Update .github/workflows/ci.yml

bushrat011899 commented:
Ahh yeah looks like I forgot to `rustup target add x86_64-unknown-none` in the runner.

cwfitzgerald commented:
Nah, I led you astray- I thought bare metal x86_64 was a tier 3 Target and therefore you actually can't tell it to install the target. This is what we have to do for vision OS and why there are two separate tool chain installation steps, but I now realize is one for tier 1 and 2 and one for tier 3

cwfitzgerald commented:
I'm going to touch up the CI script to make the definitions more clear when I do the final proper pass on this PR

cwfitzgerald committed:
CI Touchups

cwfitzgerald commented:
Thanks for all the work!

bushrat011899 commented:
Thanks for the assistance in getting it over the line! I'm planning on looking into `no_std` for `naga` next, but I suspect that'll be a much longer process.
